lib/Dialect/Arc/ArcDialect.cpp

```diff
@@ -10,16 +10,52 @@
 #include "circt/Dialect/Arc/ArcOps.h"
 #include "circt/Dialect/Arc/ArcTypes.h"
 #include "circt/Dialect/HW/HWOps.h"
+#include "mlir/Transforms/InliningUtils.h"
 
 using namespace circt;
 using namespace arc;
 
+namespace {
+/// This class defines the interface for handling inlining with Arc operations.
+struct ArcInlinerInterface : public mlir::DialectInlinerInterface {
+  using mlir::DialectInlinerInterface::DialectInlinerInterface;
+
+  bool isLegalToInline(Operation *call, Operation *callable,
+                       bool wouldBeCloned) const override {
+    if (auto stateOp = dyn_cast<StateOp>(call);
+        stateOp && stateOp.getLatency() == 0)
+      return true;
+
+    return isa<CallOp>(call);
+  }
+  bool isLegalToInline(Region *dest, Region *src, bool wouldBeCloned,
+                       IRMapping &valueMapping) const override {
+    // TODO
+    return false;
+  }
+  bool isLegalToInline(Operation *op, Region *dest, bool wouldBeCloned,
+                       IRMapping &valueMapping) const override {
+    // TODO
+    return false;
+  }
+  void handleTerminator(Operation *op,
+                        ArrayRef<Value> valuesToRepl) const override {
+    assert(isa<arc::OutputOp>(op)); // arc does not have another terminator op
+    for (auto [from, to] : llvm::zip(valuesToRepl, op->getOperands()))
+      from.replaceAllUsesWith(to);
+  }
+};
+} // end anonymous namespace
+
 void ArcDialect::initialize() {
   registerTypes();
   addOperations<
 #define GET_OP_LIST
 #include "circt/Dialect/Arc/Arc.cpp.inc"
       >();
+
+  // Register interface implementations.
+  addInterfaces<ArcInlinerInterface>();
```

Comment:
Is this enough for the MLIR inliner to pick up? How does this interact with the simple `ArcInliner` further down?

Reply:
Yes, the MLIR inliner checks for this interface and uses the overridden functions to do the inlining decisions and rewiring.
The `ArcInliner` is inheriting from a `DialectInterfaceCollection` which collects the inlining interface from all loaded (note that there is a difference between loaded and registered) dialects. The functions implemented directly in `ArcInliner` override these collected dialect specific implementations, but for functions not implemented there directly it falls back to the ones it can collect from dialects. That's why the handleTerminator function does not need to be implemented here anymore.

Reply:
Cool thanks for the clarification, that's very handy!